feat(ble): Support peripheral battery levels.

petejohanson opened the pull request:
* Add ability to fetch and report peripheral battery levels on split centrals.
* Add additional support for adding a new Battery Level service to split centrals that exposes fetched peripheral battery levels to connected hosts.

This is a continuation of the great work from @Katona in #1243 with some important changes:

* Move both central fetching and reporting of peripheral battery levels behind new (documented) Kconfig flags
* After reviewing the Battery [Level Service specification](https://www.bluetooth.com/specifications/specs/battery-service/), it became clear to me there could be an easier structure for this. In particular, that spec lays out the requirements if you're going to have more that one battery level service, how to tag a particular battery level characteristic as the main one, etc. Given that, this PR:
  * Simplifies the *core* battery reporting by simply leveraging the normal Zephyr BAS directly still, basically everywhere, including splits of both roles. I have an open PR to Zephyr to properly tag the Zephyr BAS as the main one: https://github.com/zephyrproject-rtos/zephyr/pull/65944 I am monitoring the response there, assuming no major concerns I will be cherry picking that one small change into our Zephyr fork.
  * Behind the extra Kconfig flag there is now  a dedicated new "central BAS proxy" that adds *another* Battery Level service where the Battery Level characteristics are all marked with a description of `auxiliary` to disambiguate them from the main one. The new service properly handles multiple peripherals now as well.
* Tweaks one default Zephyr L2CAP buffer value on splits to be sure we don't get `-ENOMEM` errors on reconnects to peripherals. 

Many thanks to @Katona for all the work they put into this!

# Testing

I've performed the following testing so far:

* Confirmed the battery reporting for the split central `main` battery is the one displayed on the follow OSes:
  * Linux (GNOME, using upower, on latest Fedora release)
  * macOS
  * Windows 11
  * Android 14
* Confirmed that using `bluetoothctl` I can subscribe to notifications on the *other* Battery Level service on a split central, and properly receive notifications it has "proxied" from the split peripheral.

The following testing would be nice:

* iOS device
* The custom macOS app/service that @Katona wrote.

To test, you need to enable the following off-by-default Kconfig settings:

```
CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_FETCHING=y
CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_PROXY=y
```

equiman commented:
I've merged these changes into a branch and config the `west.yml` file to use it:

```
manifest:
  remotes:
    - name: deinsoftware
      url-base: https://github.com/deinsoftware
  projects:
    - name: zmk
      remote: deinsoftware
      revision: main
      import: app/west.yml
  self:
    path: config
 ```

And tested it with the @Maksim-Isakau [zmk-split-battery](https://github.com/Maksim-Isakau/zmk-split-battery) app for windows. But it only shows one battery level.

![image](https://github.com/zmkfirmware/zmk/assets/933393/42229169-db6b-49f2-b84b-4fb503efa8a1)

Will it need some extra configuration on my [corne-zmk-config](https://github.com/deintech/corne-zmk-config)?

petejohanson commented:
> I've merged these changes into a branch and config the `west.yml` file to use it:
> 
> ```
> manifest:
>   remotes:
>     - name: deinsoftware
>       url-base: https://github.com/deinsoftware
>   projects:
>     - name: zmk
>       remote: deinsoftware
>       revision: main
>       import: app/west.yml
>   self:
>     path: config
>  ```
> 
> And tested it with the @Maksim-Isakau [zmk-split-battery](https://github.com/Maksim-Isakau/zmk-split-battery) app for windows. But it only shows one battery level.
> 
> ![image](https://github.com/zmkfirmware/zmk/assets/933393/42229169-db6b-49f2-b84b-4fb503efa8a1)

It seems that the app from @Katona assumes it just needs to use the first Battery Service, I stead of aggregating all of them from the device: https://github.com/Katona/zmk-ble/blob/main/zmk-ble/ZmkPeripheral.swift#L53

It will need to get updated to work properly.

Maksim-Isakau commented:
> And tested it with the @Maksim-Isakau [zmk-split-battery](https://github.com/Maksim-Isakau/zmk-split-battery) app for windows. But it only shows one battery level.

@equiman fixed. Please note that you have to specify
```
CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_FETCHING=y
CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_PROXY=y
```
in your KConfig file of zmk-config, split battery reporting is off by default. I stumbled upon this at first

Maksim-Isakau commented:
> iOS device

@petejohanson iOS 17 seems to work fine:
![ios-zmk-split-batt-lvl](https://github.com/zmkfirmware/zmk/assets/1701576/751e1848-828b-4d08-ab57-9d464d30edf2)

15% is peripheral, 49 main.
On a Lock Screen i see the one that has minimal battery level (tested after charging peripheral).

petejohanson commented:
> > iOS device
> 
> @petejohanson iOS 17 seems to work fine: ![ios-zmk-split-batt-lvl](https://private-user-images.githubusercontent.com/1701576/286896979-751e1848-828b-4d08-ab57-9d464d30edf2.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzNjQwMTMsIm5iZiI6MTcwMTM2MzcxMywicGF0aCI6Ii8xNzAxNTc2LzI4Njg5Njk3OS03NTFlMTg0OC04MjhiLTRkMDgtYWI1Ny05ZDQ2NGQzMGVkZjIuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTEzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzExMzBUMTcwMTUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTJiNzcwMDM3ODM0ZjAzMTEzMDZmNjhmNTZmZDQ0Njc3ZjhiZmFkMTY1YjQ2MTMyODJhOTAwOTg3YmRmZmE3MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Jocx4kaY6D4htSkKxHXsp1OBzWbP2ZbC7jGeXbmQprw)
> 
> 15% is peripheral, 49 main. On a Lock Screen i see the one that has minimal battery level (tested after charging peripheral).

It's a shame they don't use our display string at all or otherwise disambiguate the two battery levels, but at least they show them both.

petejohanson commented:
> Thanks @petejohanson  and @Maksim-Isakau. Now it's working again.
> 
> ![image](https://github.com/zmkfirmware/zmk/assets/933393/ea7ef5ef-0dc1-4ef1-9280-3ca740f78267)
> 
> It's possible to add some config labels to rename each one?
> 
> I think `Peripheral #: ` is descriptive but not a cool name,
> I would like to rename it as `Left` and `Right` for example.
> 

I wonder if perhaps I can enhance this to add a CUD based on the keyboard name? Or we offer an additional Kconfig for the battery  label, and we proxy the CUD from the connected peripherals once they are connected....

petejohanson commented:
Ok, so the Zephyr PR was accepted and merged, so I've cherry picked that into our Zephyr fork in https://github.com/zmkfirmware/zephyr/pull/27

Waiting to see if @Katona is able to work on his macOS app bits, but I don't think that's a blocker for merging this in the meantime.

petejohanson commented:
> Well, today something strange is happening. It's showing all battery levels at -1
> 
> ![image](https://private-user-images.githubusercontent.com/933393/287340665-b4ac9111-981e-4df0-8b2a-f3572bbf6b47.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE0NTcxODYsIm5iZiI6MTcwMTQ1Njg4NiwicGF0aCI6Ii85MzMzOTMvMjg3MzQwNjY1LWI0YWM5MTExLTk4MWUtNGRmMC04YjJhLWYzNTcyYmJmNmI0Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwMVQxODU0NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZmMzNmVmM2QxNGMyYzhmYWZhODZmYmVmNmMzNDIzNDA2ZGVlZDI2NzkxMTcxOTNiOWE5ODkyZTczNTQ4MDlmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.87uUSHkOoyVLzYH5nYSDRm-I4h2TkrDqCa8PCimor4c)
> 
> Or sometimes only, show the central level:
> 
> ![image](https://private-user-images.githubusercontent.com/933393/287340777-78af19e6-779a-4ffa-b980-81111c68849d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE0NTcxODYsIm5iZiI6MTcwMTQ1Njg4NiwicGF0aCI6Ii85MzMzOTMvMjg3MzQwNzc3LTc4YWYxOWU2LTc3OWEtNGZmYS1iOTgwLTgxMTExYzY4ODQ5ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwMVQxODU0NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NDMxNjBmY2VhMzMyNGU3MDk4ZjZkNmM2YWNiNGM4MzhiYTFmZTJiNGQ1NjM1OGEzYjU1MjdkMmFkYzQ1OWRkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.KPG7Jspe9QNzFe-PC4MrJEDSJSBPX5UX8nRVtX6o2UY)

Is the main battery display in Windows working?

equiman commented:
Yes, it is: 
![image](https://github.com/zmkfirmware/zmk/assets/933393/102cc1f2-ead8-4222-853b-6ee2cd42ad19)

![image](https://github.com/zmkfirmware/zmk/assets/933393/0a6f5c4e-184f-4d6a-889f-b46ca3d9c4f0)

bryanforbes commented:
It seems like there should be an API for getting the peripheral battery level that's similar to `zmk_battery_state_of_charge()`. This would be useful for displaying peripheral battery level in the RGB underglow like the Glove80 does. Right now, the only place where the levels are stored is in the proxy, but it seems wrong to force users to enable the proxy just to get the levels for use elsewhere. Which means the simplest solution would be to store the levels in `central.c` when fetching is enabled. (cc @chrisandreae)

petejohanson commented:
> It seems like there should be an API for getting the peripheral battery level that's similar to `zmk_battery_state_of_charge()`. This would be useful for displaying peripheral battery level in the RGB underglow like the Glove80 does. Right now, the only place where the levels are stored is in the proxy, but it seems wrong to force users to enable the proxy just to get the levels for use elsewhere. Which means the simplest solution would be to store the levels in `central.c` when fetching is enabled. (cc @chrisandreae)

Changed the storage to the central code, added a new API for fetching the levels there, and then leveraged that in the BAS proxy code instead.

equiman commented:
Few days ago, the keyboard stop reporting correctly the battery level, even on windows. 

Always shown the level in 50%. Today it was run on battery, and report it has 50%. Once I've fully charged it, still reporting 50%.

![image](https://github.com/zmkfirmware/zmk/assets/933393/f226fdca-5f48-4b07-be09-9cf654d6e264)

And the battery status won't connect any more:

![image](https://github.com/zmkfirmware/zmk/assets/933393/611a9c97-7460-4d61-82b5-a14007b8e59d)

petejohanson commented:
> Few days ago, the keyboard stop reporting correctly the battery level, even on windows.
> 
> Always shown the level in 50%. Today it was run on battery, and report it has 50%. Once I've fully charged it, still reporting 50%.
> 
> ![image](https://private-user-images.githubusercontent.com/933393/289313918-f226fdca-5f48-4b07-be09-9cf654d6e264.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIxNTkxMDksIm5iZiI6MTcwMjE1ODgwOSwicGF0aCI6Ii85MzMzOTMvMjg5MzEzOTE4LWYyMjZmZGNhLTVmNDgtNGIwNy1iZTA5LTljZjY1NGQ2ZTI2NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwOVQyMTUzMjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNDc4OTEzNTEwY2I3YTNiZDAzNzE0NTI4YzIyYjQ0ODRmODI4ZTE0MGVhYWEzODJjOGZiOTdhZDNjYzhlNTE5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.REkkl2wv_KhZD3bysjCcW05AXXQPq32XROxeauzw3MM)
> 
> And the battery status won't connect any more:
> 
> ![image](https://private-user-images.githubusercontent.com/933393/289314186-611a9c97-7460-4d61-82b5-a14007b8e59d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIxNTkxMDksIm5iZiI6MTcwMjE1ODgwOSwicGF0aCI6Ii85MzMzOTMvMjg5MzE0MTg2LTYxMWE5Yzk3LTc0NjAtNGQ2MS04MmI1LWExNDAwN2I4ZTU5ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwOVQyMTUzMjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xZWJjMDI1ZTM0YjA4ZmVjMmRhMTBmZTRjZmMzOWI2YjI4ZTU5YjVhMWEyYmMyMmExYTBjMWY1MDdmMWQ0Y2MwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.gieTW8NiuUGtVwTqpN5MSKm6k5BQwOkHO86WD3BewjQ)

Interesting... Can you please try the latest code I just pushed? Also, getting the ZMK logs when this happens would be useful. Thanks for testing!

caksoylar commented on docs/docs/config/system.md:
```diff
@@ -104,15 +104,18 @@ Note that `CONFIG_BT_MAX_CONN` and `CONFIG_BT_MAX_PAIRED` should be set to the s
-| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_STACK_SIZE`          | int  | Stack size of the BLE split peripheral notify thread                     | 650     |
-| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_PRIORITY`            | int  | Priority of the BLE split peripheral notify thread                       | 5       |
-| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_POSITION_QUEUE_SIZE` | int  | Max number of key state events to queue to send to the central           | 10      |
+| Config                                                  | Type | Description                                                                | Default                                    |
+| ------------------------------------------------------- | ---- | -------------------------------------------------------------------------- | ------------------------------------------ |
+| `CONFIG_ZMK_SPLIT`                                      | bool | Enable split keyboard support                                              | n                                          |
+| `CONFIG_ZMK_SPLIT_PERIPHERAL_HID_INDICATORS`            | bool | Enable split keyboard support for passing indicator state to peripherals   | n                                          |
+| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_FETCHING`   | bool | Enable fetching split peripheral battery levels to the central side        | n                                          |
+| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_PROXY`      | bool | Enable central reporting of split battery levels to hosts                  | n                                          |
+| `CONFIG_ZMK_SPLIT_BLE`                                  | bool | Use BLE to communicate between split keyboard halves                       | y                                          |
+| `CONFIG_ZMK_SPLIT_ROLE_CENTRAL`                         | bool | `y` for central device, `n` for peripheral                                 |                                            |
+| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_QUEUE_SIZE` | int  | Max number of battery level events to queue when received from peripherals | `CONFIG_ZMK_SPLIT_BLE_CENTRAL_PERIPHERALS` |
+| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_POSITION_QUEUE_SIZE`      | int  | Max number of key state events to queue when received from peripherals     | 5                                          |
+| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_SPLIT_RUN_STACK_SIZE`     | int  | Stack size of the BLE split central write thread                           | 512                                        |
+| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_SPLIT_RUN_QUEUE_SIZE`     | int  | Max number of behavior run events to queue to send to the peripheral(s)    | 5                                          |
+| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_STACK_SIZE`            | int  | Stack size of the BLE split peripheral notify thread                       | 650                                        |
+| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_PRIORITY`              | int  | Priority of the BLE split peripheral notify thread                         | 5                                          |
+| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_POSITION_QUEUE_SIZE`   | int  | Max number of key state events to queue to send to the central             | 10                                         |
```
```suggestion
| Config                                                  | Type | Description                                                                | Default                                    |
| ------------------------------------------------------- | ---- | -------------------------------------------------------------------------- | ------------------------------------------ |
| `CONFIG_ZMK_SPLIT`                                      | bool | Enable split keyboard support                                              | n                                          |
| `CONFIG_ZMK_SPLIT_PERIPHERAL_HID_INDICATORS`            | bool | Enable split keyboard support for passing indicator state to peripherals   | n                                          |
| `CONFIG_ZMK_SPLIT_ROLE_CENTRAL`                         | bool | `y` for central device, `n` for peripheral                                 |                                            |
| `CONFIG_ZMK_SPLIT_BLE`                                  | bool | Use BLE to communicate between split keyboard halves                       | y                                          |
| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_FETCHING`   | bool | Enable fetching split peripheral battery levels to the central side        | n                                          |
| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_PROXY`      | bool | Enable central reporting of split battery levels to hosts                  | n                                          |
| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_BATTERY_LEVEL_QUEUE_SIZE` | int  | Max number of battery level events to queue when received from peripherals | `CONFIG_ZMK_SPLIT_BLE_CENTRAL_PERIPHERALS` |
| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_POSITION_QUEUE_SIZE`      | int  | Max number of key state events to queue when received from peripherals     | 5                                          |
| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_SPLIT_RUN_STACK_SIZE`     | int  | Stack size of the BLE split central write thread                           | 512                                        |
| `CONFIG_ZMK_SPLIT_BLE_CENTRAL_SPLIT_RUN_QUEUE_SIZE`     | int  | Max number of behavior run events to queue to send to the peripheral(s)    | 5                                          |
| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_STACK_SIZE`            | int  | Stack size of the BLE split peripheral notify thread                       | 650                                        |
| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_PRIORITY`              | int  | Priority of the BLE split peripheral notify thread                         | 5                                          |
| `CONFIG_ZMK_SPLIT_BLE_PERIPHERAL_POSITION_QUEUE_SIZE`   | int  | Max number of key state events to queue to send to the central             | 10                                         |
```
This ordering (moving `CONFIG_ZMK_SPLIT_ROLE_CENTRAL` up and then `CONFIG_ZMK_SPLIT_BLE` before BLE-specific options) makes more sense to me, but not a big deal.

petejohanson replied:
Fair. Tweaked the ordering.

caksoylar reviewed (approved):
Docs look good to me, noted one nit re: table ordering.

equiman commented:
> Interesting... Can you please try the latest code I just pushed? Also, getting the ZMK logs when this happens would be useful. Thanks for testing!

Sure, I'll try to do it once I finish something.
About the logs, is there some link or documentation about how to enable it and where the logs files are stored?

petejohanson commented:
> 
> Sure, I'll try to do it once I finish something.
> About the logs, is there some link or documentation about how to enable it and where the logs files are stored?  

https://zmk.dev/docs/development/usb-logging

bryanforbes commented:
I just tested this with my Glove80 on my MacBook Pro running Sonoma with the battery widget and both halves report their levels properly:

<img width="364" alt="image" src="https://github.com/zmkfirmware/zmk/assets/204106/13fdd694-8fe3-40a7-aa53-54958c8f8ea6">

I was also able to update the Glove80's underglow status to use the new API for fetching the peripheral levels successfully.

equiman commented:
> Interesting... Can you please try the latest code I just pushed?

I've merged the last changes and it's working again 🎉🥳

![image](https://github.com/zmkfirmware/zmk/assets/933393/3bf92cee-09f1-47b1-9ad3-89d273fcdb9e)

joelspadin commented on app/src/split/bluetooth/central.c:
```diff
@@ -265,6 +270,95 @@ static uint8_t split_central_notify_func(struct bt_conn *conn,
+static uint8_t peripheral_battery_levels[ZMK_SPLIT_BLE_PERIPHERAL_COUNT] = {0};
+
+int zmk_split_get_peripheral_battery_level(uint8_t source, uint8_t *level) {
+    if (source < 0 || source >= ZMK_SPLIT_BLE_PERIPHERAL_COUNT) {
```
`source` is unsigned, so it can't be less than 0.

Also, consider using `ARRAY_SIZE(peripheral_battery_levels)` so the bounds check stays in sync with the size of the array it's checking, even if that size is changed to use a different symbol later.

petejohanson replied:
Good call. changed.

joelspadin commented on app/src/split/bluetooth/central.c:
```diff
@@ -265,6 +270,95 @@ static uint8_t split_central_notify_func(struct bt_conn *conn,
+                                                       const void *data, uint16_t length) {
+    struct peripheral_slot *slot = peripheral_slot_for_conn(conn);
+
+    if (slot == NULL) {
```
Minor: seems a bit inconsistent to use `slot == NULL` for one null check and `!data` for another.

petejohanson replied:
Made these consistent.

joelspadin commented on app/src/split/bluetooth/central.c:
```diff
@@ -265,6 +270,95 @@ static uint8_t split_central_notify_func(struct bt_conn *conn,
+    }
+
+    LOG_DBG("[BATTERY LEVEL NOTIFICATION] data %p length %u", data, length);
+    uint8_t battery_level = ((uint8_t *)data)[0];
```
Do we need to verify that `length >= 1` or is that ensured elsewhere?

petejohanson replied:
Good catch. Added checks for this in case we have an empty notification payload.

joelspadin commented on app/src/split/bluetooth/central.c:
```diff
@@ -265,6 +270,95 @@ static uint8_t split_central_notify_func(struct bt_conn *conn,
+    return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t split_central_battery_level_read_func(struct bt_conn *conn, uint8_t err,
```
Comments from the above function apply here too.

petejohanson replied:
Indeed. Fixed.

joelspadin commented on app/src/split/bluetooth/central_bas_proxy.c:
```diff
@@ -0,0 +1,92 @@
+
+    // 1 offsets for the service attribute, then offset 5 for each battery because that's how
+    // many attributes are added per battery
+    uint8_t index = 1 + (5 * ev->source);
```
You might as well just use `int` here. Not that anyone will likely try to set up something with 51 peripherals and cause this to overflow, but using an int that matches the register size can be more efficient, and the extra stack space is either zero or negligible depending on what the compiler decides to do.

petejohanson replied:
Fair. Done.

joelspadin commented on app/src/split/bluetooth/central_bas_proxy.c:
```diff
@@ -0,0 +1,92 @@
+
+    LOG_DBG("Peripheral battery level event: %u", ev->state_of_charge);
+
+    // 1 offsets for the service attribute, then offset 5 for each battery because that's how
```
>  then offset 5 for each battery because that's how many attributes are added per battery

Is there a `#define` somewhere for this, or is this just something baked into the spec and unlikely to change?

petejohanson replied:
Tweaked this for defines to hopefully makes this clearer.

chrisandreae commented on app/src/split/bluetooth/central.c:
```diff
@@ -265,6 +270,105 @@ static uint8_t split_central_notify_func(struct bt_conn *conn,
+    struct zmk_peripheral_battery_state_changed ev;
+    while (k_msgq_get(&peripheral_batt_lvl_msgq, &ev, K_NO_WAIT) == 0) {
+        LOG_DBG("Triggering peripheral battery level change %u", ev.state_of_charge);
+        peripheral_battery_levels[ev.source] = ev.state_of_charge;
```
Should the `peripheral_battery_levels` value also be reset to 0 (or a sentinel value?) when the peripheral disconnects?

Additionally, should there be a matching API to permit consumers to ask about the peripheral's presence without subscribing to `zmk_split_peripheral_status_changed` and maintaining an independent local view on the state?

A consumer of the peripheral battery level such as a display would almost certainly want to display something different when the peripheral is disconnected.

caksoylar replied:
Worth noting that https://github.com/zmkfirmware/zmk/pull/2033 adds an API for checking peripheral bonded/connected status.

chrisandreae replied:
Oh, excellent. Well, then those would play nicely, then. 👍 

In addition to that though, I think it might still still be worth changing `zmk_split_get_peripheral_battery_level` here to return an error code if the specified peripheral is not connected.

petejohanson replied:
> Oh, excellent. Well, those would play nicely, then. 👍
> 
> In addition to that though, I think it might still still be worth changing `zmk_split_get_peripheral_battery_level` here to return an error code if the specified peripheral is not connected.

Good call. Added that check and the API now returns `-ENOTCONN` in that scenario. Pushed the changes.

petejohanson replied:
Ok, I've also added code to trigger an event on disconnect to set the level back to 0 for the newly disconnected peripheral as well. I believe that rounds this out enough.

equiman commented:
I've merged the branch with the last changes, and still working on windows:

![image](https://github.com/zmkfirmware/zmk/assets/933393/85511c97-5c1a-49fd-aee7-3f0bf84f2ecb)

bryanforbes commented:
I've been using the latest for roughly a day and I noticed that it seems to be reporting the right side (peripheral) battery level of my Glove80 as the main bluetooth battery level in macOS:

<img width="352" alt="image" src="https://github.com/zmkfirmware/zmk/assets/204106/63161fb8-8ce4-4e20-90d0-602d591e6a1a">
<img width="303" alt="image" src="https://github.com/zmkfirmware/zmk/assets/204106/01fc6433-03a3-413f-aad6-08fa7e66abdc">

The battery widget (first image) is correctly ordering the halves: left on top (89%), right below it (100%). But the BT menu is reporting the battery level as 100% (right half).

petejohanson commented:
> I've been using the latest for roughly a day and I noticed that it seems to be reporting the right side (peripheral) battery level of my Glove80 as the main bluetooth battery level in macOS:
> <img alt="image" width="352" src="https://private-user-images.githubusercontent.com/204106/293293248-63161fb8-8ce4-4e20-90d0-602d591e6a1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDM4NzI5NTcsIm5iZiI6MTcwMzg3MjY1NywicGF0aCI6Ii8yMDQxMDYvMjkzMjkzMjQ4LTYzMTYxZmI4LThjZTQtNGUyMC05MGQwLTYwMmQ1OTFlNmExYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjMxMjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIyOVQxNzU3MzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNGQ1MzkwNmEyNTc4NzJlMmZjMGFjMTc3ZDEwYWE4MGExMjhhNDEzZDc4ZDQ1MzFiNmZkZWFiMDcwOTc0NGFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.uQpqr4pgksLTvt8bVNL5uXMDImHcPo0r_GfKX_Y_Mec"> <img alt="image" width="303" src="https://private-user-images.githubusercontent.com/204106/293293266-01fc6433-03a3-413f-aad6-08fa7e66abdc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDM4NzI5NTcsIm5iZiI6MTcwMzg3MjY1NywicGF0aCI6Ii8yMDQxMDYvMjkzMjkzMjY2LTAxZmM2NDMzLTAzYTMtNDEzZi1hYWQ2LTA4ZmE3ZTY2YWJkYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjMxMjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIyOVQxNzU3MzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xZWI0YmRlMjY3MmRhMzM4YmE1OGU3ZDA2ZDdiZDExYzI3MDNiZTlmNmRkNTkxZGMzZTBmNTY1ZTRmY2ZhNzFjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.1LcW4LD0-p-E0xk2haKQcwv1WjKwmniiDYtUpW6h2vg">
> 
> The battery widget (first image) is correctly ordering the halves: left on top (89%), right below it (100%). But the BT menu is reporting the battery level as 100% (right half).
> 
> Update: After a while (I'm not sure how long), the BT menu started reporting the left half percentage and the battery widget switched the order of the two halves from the images I uploaded earlier (left on bottom, right on top). Could this be some sort of race condition?

Not sure it's a race condition, as much as indeterminate behavior on how macOS handles multiple battery level characteristics split over two different battery GATT services... Initial testing seemed to show that they properly picked the characteristic marked as the `main`.

Are you perhaps on a different underlying Zephyr branch, testing the Zephyr 3.5 work? Not sure I cherry picked a needed Zephyr fix there (will be doing that momentarily)

bryanforbes commented:
> Are you perhaps on a different underlying Zephyr branch, testing the Zephyr 3.5 work?

No, I’m using MoErgo’s main (Zephyr 3.2) as my base and replaced their split battery reporting commit with the work here and updated the underglow status to use the new API. I’ll test this against a stock ZMK later this afternoon to try to narrow down the issue.

petejohanson commented:
> > Are you perhaps on a different underlying Zephyr branch, testing the Zephyr 3.5 work?
> 
> No, I’m using MoErgo’s main (Zephyr 3.2) as my base and replaced their split battery reporting commit with the work here and updated the underglow status to use the new API. I’ll test this against a stock ZMK later this afternoon to try to narrow down the issue.

I *suspect* you'll find the same thing occurring, and it's simply that the macOS selection of the battery level to display is non-deterministic.

petejohanson commented:
> I've been using the latest for roughly a day and I noticed that it seems to be reporting the right side (peripheral) battery level of my Glove80 as the main bluetooth battery level in macOS:
> <img alt="image" width="352" src="https://private-user-images.githubusercontent.com/204106/293293248-63161fb8-8ce4-4e20-90d0-602d591e6a1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyMjk3ODUsIm5iZiI6MTcwNDIyOTQ4NSwicGF0aCI6Ii8yMDQxMDYvMjkzMjkzMjQ4LTYzMTYxZmI4LThjZTQtNGUyMC05MGQwLTYwMmQ1OTFlNmExYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwMlQyMTA0NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZDEzNDIxZmZmM2ZmYTczYzQyMzNjZWI0NzYzZDNjMWM2ZDdjM2U2YmJiY2NjNDRjMGI3NTc4MzhmNjExOWYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.iiDeHyjegg4h70CY-zDxlQEWUvEkHqqJgB2yATE5rxs"> <img alt="image" width="303" src="https://private-user-images.githubusercontent.com/204106/293293266-01fc6433-03a3-413f-aad6-08fa7e66abdc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyMjk3ODUsIm5iZiI6MTcwNDIyOTQ4NSwicGF0aCI6Ii8yMDQxMDYvMjkzMjkzMjY2LTAxZmM2NDMzLTAzYTMtNDEzZi1hYWQ2LTA4ZmE3ZTY2YWJkYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwMlQyMTA0NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03Zjg5ZjVmMmFhMDVmN2Q2ZGQ2Zjg0NjAwYTI1NmI1ZDkzMjNjNDg5NjM2ZjNiODMwYjcwNjI4YzlmMjZjZmQxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.VdgnPnFN6onnuTjXVxX4RVTir5os87bOPBVuKc6qjPE">
> 
> The battery widget (first image) is correctly ordering the halves: left on top (89%), right below it (100%). But the BT menu is reporting the battery level as 100% (right half).
> 
> Update: After a while (I'm not sure how long), the BT menu started reporting the left half percentage and the battery widget switched the order of the two halves from the images I uploaded earlier (left on bottom, right on top). Could this be some sort of race condition?

I don't think we're getting any help any time soon...

![image](https://github.com/zmkfirmware/zmk/assets/202695/86f61b0b-c9d0-443c-9768-e6d46fcb35d1)

My gut is to not hold up our merge of this just because macOS doesn't do sane battery level characteristic selection. The new Sonoma battery widget actually shows the multiple levels. *Hopefully* it will do so with labels when we add the GATT CUD to each battery level.

Also, the proxy portion is optional, and not enabled by default so there's less risk of this being an issue while we continue to iterate.

@bryanforbes Any concerns?

bryanforbes commented:
No concerns from me

dudicoco commented:
I've added the parameters to my config but I do not see peripheral battery status on macOs:
 
![Screenshot 2024-01-05 at 17 11 29](https://github.com/zmkfirmware/zmk/assets/51051860/a7655ff1-0256-46e7-b4f1-0ef607bf8715)

https://github.com/dudicoco/corne-wireless-zmk-config/blob/dacb95898f9254b86278ddb7112ca28b014f9655/config/corne.conf#L20-L23

madushan1000 commented:
here is a python script to see split battery levels on linux. https://gist.github.com/madushan1000/9744eb6350a5dd9685fb6bfbb25fbb8a

dstudzinski commented:
> I've been using the latest for roughly a day and I noticed that it seems to be reporting the right side (peripheral) battery level of my Glove80 as the main bluetooth battery level in macOS:
> 
> 
> 
> <img width="352" alt="image" src="https://github.com/zmkfirmware/zmk/assets/204106/63161fb8-8ce4-4e20-90d0-602d591e6a1a">
> 
> <img width="303" alt="image" src="https://github.com/zmkfirmware/zmk/assets/204106/01fc6433-03a3-413f-aad6-08fa7e66abdc">
> 
> 
> 
> The battery widget (first image) is correctly ordering the halves: left on top (89%), right below it (100%). But the BT menu is reporting the battery level as 100% (right half).
> 
> 
> 
> Update: After a while (I'm not sure how long), the BT menu started reporting the left half percentage and the battery widget switched the order of the two halves from the images I uploaded earlier (left on bottom, right on top). Could this be some sort of race condition?

What's your widget/app for macos? I can't find any.

bryanforbes commented:
> What's your widget/app for macos? I can't find any.

It’s just the battery widget in Sonoma in the notification center.

bjornborg commented:
> here is a python script to see split battery levels on linux. https://gist.github.com/madushan1000/9744eb6350a5dd9685fb6bfbb25fbb8a

Great script. I have fused this to another gnome extension, so it is graphically shown in gnome.
![image](https://github.com/zmkfirmware/zmk/assets/9810017/4fa6d580-3e35-461e-97fd-cf2b1e001a6f)
The repo is located here: https://github.com/bjornborg/gnome-bluetooth-battery-indicator/tree/gnome.42

mh4x0f commented:
I created an app inspired by the @Maksim-Isakau  project just for Linux on Dbus Bluez.

![image](https://github.com/zmkfirmware/zmk/assets/6795613/05187487-2eab-457c-ae60-6da50482279c)


https://github.com/mh4x0f/zmkBATx

dudicoco commented:
> I've been using the latest for roughly a day and I noticed that it seems to be reporting the right side (peripheral) battery level of my Glove80 as the main bluetooth battery level in macOS:
> 
> <img alt="image" width="352" src="https://private-user-images.githubusercontent.com/204106/293293248-63161fb8-8ce4-4e20-90d0-602d591e6a1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA4ODAyMzgsIm5iZiI6MTcxMDg3OTkzOCwicGF0aCI6Ii8yMDQxMDYvMjkzMjkzMjQ4LTYzMTYxZmI4LThjZTQtNGUyMC05MGQwLTYwMmQ1OTFlNmExYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMxOVQyMDI1MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZmVmNTA4M2RlZDQyYWM2YzJhYjg4NDFkNTdlY2NmNTA4OThkYmFhYmVhNzc1NDYzOTNmMTFjZTg3MTgzYTA1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.MffJuHZSk0wPvHnKdpCMr87ADlY-Bt9OQ3k55FjmLTw"> <img alt="image" width="303" src="https://private-user-images.githubusercontent.com/204106/293293266-01fc6433-03a3-413f-aad6-08fa7e66abdc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA4ODAyMzgsIm5iZiI6MTcxMDg3OTkzOCwicGF0aCI6Ii8yMDQxMDYvMjkzMjkzMjY2LTAxZmM2NDMzLTAzYTMtNDEzZi1hYWQ2LTA4ZmE3ZTY2YWJkYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMxOVQyMDI1MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02NTVjYWJjNGE0NGUwZmZmNDI3Yjc2YzVjNGRmODEyZTc3NDg0M2RmZWZiOGU3Y2M2YThlNmNmOTM3NTdlMTc5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ayCQ4LOG24wSyrXwqE6dYZsiYbEjOZFPYqusfV4XQsY">
> The battery widget (first image) is correctly ordering the halves: left on top (89%), right below it (100%). But the BT menu is reporting the battery level as 100% (right half).
> 
> Update: After a while (I'm not sure how long), the BT menu started reporting the left half percentage and the battery widget switched the order of the two halves from the images I uploaded earlier (left on bottom, right on top). Could this be some sort of race condition?

@bryanforbes can you please share what widget are you using for mac?

dudicoco commented:
Figured out how to display the split battery status on mac:
1. Follow the instructions on how to add a widget - https://support.apple.com/en-il/guide/mac-help/mchl52be5da5/mac.
2. Add the battery widget.

mike1808 commented:
> here is a python script to see split battery levels on linux. https://gist.github.com/madushan1000/9744eb6350a5dd9685fb6bfbb25fbb8a

thanks! created a polybar module if anyone uses it with i3 or other wms

https://gist.github.com/mike1808/13af77145a292c627fa6508f5fb6730b
![image](https://github.com/zmkfirmware/zmk/assets/3942554/068afd20-6dc2-4551-b08b-a17f5e1faae6)

rottencandy commented:
> here is a python script to see split battery levels on linux. https://gist.github.com/madushan1000/9744eb6350a5dd9685fb6bfbb25fbb8a

Thank you, I made a bash version of this script
https://gist.github.com/rottencandy/901189a3cfac94bcbdefdf40236b042e

linkarzu commented:
@bryanforbes or any macOS user out there, has someone been able to get the battery values via CLI?
The reason I'm asking, is because I want to get the values, and show them in SketchyBar and probably set a notification when battery is around 5% so that I charge the keyboard

As seen on image2, I do see the values in the macOS battery widget, but I'd like to get it via a terminal command

<div align="left">
    <img src="https://github.com/user-attachments/assets/788b4311-e0e3-4ef0-b6ce-e66903fd7dfa" width="600">
</div>

<div align="left">
    <img src="https://github.com/user-attachments/assets/b3c402b7-6b9f-4726-9ceb-8b15aead8b74" width="250">
</div>
